cedar-lean/Cedar/Validation/RequestEntityValidator.lean

```diff
@@ -0,0 +1,157 @@
+import Cedar.Validation.Validator
+import Cedar.Validation.Typechecker
+namespace Cedar.Validation
+
+open Cedar.Spec
+open Cedar.Data
+
+inductive RequestValidationError where
+| typeError (loc : String)
+
+abbrev RequestValidationResult := Except RequestValidationError Unit
+
+inductive EntityValidationError where
+| typeError (loc : String)
+
+abbrev EntityValidationResult := Except EntityValidationError Unit
+
+def instanceOfBoolType (b : Bool) (bty : BoolType) : Bool :=
+match b, bty with
+  | true, .tt => true
+  | false, .ff => true
+  | _, .anyBool => true
+  | _, _ => false
+
+def instanceOfEntityType (e : EntityUID) (ety : EntityType ) : Bool := ety == e.ty
+
+def instanceOfExtType (ext : Ext) (extty: ExtType) : Bool :=
+match ext, extty with
+  | .decimal _, .decimal => true
+  | .ipaddr _, .ipAddr => true
+  | _, _ => false
+
+def requiredAttributesPresent (r : Map Attr Value) (rty : Map Attr (Qualified CedarType)) (k : Attr) :=
+match rty.find? k with
+    | .some qty => if qty.isRequired then r.contains k else true
+    | _ => true
+
+def instanceOfType (v : Value) (ty : CedarType) : Bool := match v, ty with
+| .prim (.bool b), .bool bty => instanceOfBoolType b bty
+| .prim (.int _), .int => true
+| .prim (.string _), .string => true
+| .prim (.entityUID e), .entity ety => instanceOfEntityType e ety
+| .set s, .set ty => (s.elts.attach.map (λ ⟨v, _⟩ => instanceOfType v ty)).all id
+| .record r, .record rty =>
+  r.keys.all rty.keys.contains &&
+  (r.kvs.attach₂.all (λ ⟨(k, v), _⟩ => (match rty.find? k with
+      | .some qty => instanceOfType v qty.getType
+      | _ => true))) &&
+  rty.keys.all (requiredAttributesPresent r rty)
+| .ext x, .ext xty => instanceOfExtType x xty
+| _, _ => false
+  termination_by v
+  decreasing_by
+    all_goals simp_wf
+    case _ h₁ =>
+      have := Set.sizeOf_lt_of_mem h₁
+      omega
+    case _ h₁ =>
+      cases r
+      simp only [Map.kvs] at h₁
+      simp only [Map.mk.sizeOf_spec]
+      omega
+
+def instanceOfRequestType (request : Request) (reqty : RequestType) : Bool :=
```

Comment:
Why use nested if-then-else here?  The following should be equivalent:

```
instanceOfEntityType request.principal reqty.principal &&
request.action == reqty.action && 
instanceOfEntityType request.resource reqty.resource &&
instanceOfType request.context (.record reqty.context)

```

Reply:
ah it was nested because i was returning a custom error for each one before; i changed it in the last pass